### ISSUE TITLE
Wrapper for the ginkgo numerical linear algebra library.

### ISSUE THREAD

**Issue by pratikvn** [dealii/dealii#7183]
Hello,

I would like to add a wrapper for the [ginkgo](https://github.com/ginkgo-project/ginkgo) library within deal.ii. 

I have forked the repository [within the ginkgo project](https://github.com/ginkgo-project/dealii). Currently only the automatic detection of Ginkgo within deal.ii's has been added. Do you recommend a pull request for this as well? or maybe a pull request for a larger addition ?

The next steps would be:

1. An interface for Ginkgo's sparse matrix, so that there are no unnecessary copies. 
A question: Would you also like an option to be able to select the different matrix formats, for example selection within CG so that solution with COO or ELL or other formats would be possible. I see this being set within the ``AdditionalData`` struct passed to the solver.

2. Addition of the ``SolverControl`` interface mainly for stopping criteria of the iterative solvers.

2. Addition of the solver interface and the various iterative solvers.

3. Addition of the preconditioner interface and the different preconditioners.

4. Maybe also some logging ? (Depending on what you need)

Please feel free to go through the source code of [ginkgo](https://github.com/ginkgo-project/ginkgo) and if you have any recommendations, please let us know.

**Comment by drwells** [comment]
Ginkgo looks interesting. Is there a whitepaper on it? I was not able to find one linked to in the repository.

If the goal is to just use Ginkgo matrices in deal.II's iterative solvers (`SolverCG` and friends) then all they need to do is implement `vmult`, `vmult_add`, and possibly `Tvmult` and `Tvmult_add`. I suspect that you could use `LinearOperator` to wrap a Ginkgo matrix relatively easily.

I am not sure how useful things like COO or other sparse matrix formats are in this context: AFAIK all deal.II applications that use matrices start by computing the sparsity pattern anyway, which matches the way we handle constraints and assembly.

If you wanted to add a more involved interface (i.e., support element access) I would take a look at our present CUDA, Trilinos, or PETSc wrappers; those wrap other matrix classes in a way that is compatible with the rest of the library.

**Comment by pratikvn** [comment]
Thanks. [Here](https://github.com/ginkgo-project/outreach) is some material explaining some ideas behind ginkgo.

Yes, we aim for a more involved interface but add the linear solvers from Ginkgo. We do not want to add the matrix formats to deal.ii. What I meant to say was if you would like to be able to pass as an option to Ginkgo's linear solvers, the different matrix formats. The matrix formats would be used only within ginkgo but the option to choose them would be from deal.ii. You might get quite some speedup for solution times if you can change the matrix formats and use the one that is suitable for your matrix.

You can find a draft of the white paper attached which gives a more detailed explanation of the ideas and implementation involved. [TOMS_unfinished.pdf](https://github.com/dealii/dealii/files/2382910/TOMS_unfinished.pdf)

For this paper, we would also like to include the interface to deal.ii.

**Comment by drwells** [comment]
I read through the preprint and I think I have a better idea now of what Ginkgo offers in terms of execution strategies and why its own solvers should be used.

It would be interesting to see how Ginkgo compares to something like Kokkos or our own TBB-based sparse matrix support.

> We do not want to add the matrix formats to deal.ii. What I meant to say was if you would like to be able to pass as an option to Ginkgo's linear solvers, the different matrix formats.

I am not sure that I understand what you are proposing: we need to add the matrix formats to deal.II to use things like `AffineConstraints` and other matrix manipulation code. Otherwise we will have to copy the matrix right before we use the Ginkgo solver. Do you mean that we should go the other way, i.e., make Ginkgo aware of deal.II's default `SparseMatrix` format?

Of course, if all you are interested in doing is a performance comparison, then I would just copy things with matrix iterators to a Ginkgo object and then use a Ginkgo solver: that would remove the need for most of the glue code.

**Comment by bangerth** [comment]
I am positively convinced that copying the matrix once from deal.II to Gingko format at the beginning of a solver step has no impact on performance. Compared to the work that needs to be done in a solver (as well as the work necessary to assemble it in the first place), copying once won't make a difference.

**Comment by pratikvn** [comment]
Currently as we do not have MPI support, I think would not make sense to use ginkgo's matrix as the main matrix for all of deal.ii's operations. What I was thinking of was more in the direction of just borrowing the pointers from deal.ii's system matrix after it has been assembled, basically getting the raw pointers to the colidxs, rowptrs and val_ptrs (as your default is CRS) and creating a ginkgo matrix from these pointers and using them to solve using ginkgo's solvers. 

Yes, I understand that considering assembly time and the actual solution time, the time taken for copies would probably not make a difference. But as it is quite easy to prevent these copies by just declaring Ginkgo's sparse matrix class (a wrapper in deal.ii) a friend of the actual SparseMatrix I think it is doable. Of course, if you dont want outer libraries accessing the raw pointers or if you really prefer that we copy nonetheless, I am happy to do that as well. 

So, in short, I dont want to add Ginkgo's matrix formats to deal.ii (atleast for now) because we dont want to manipulate any raw matrix data within deal.ii. Ideally, deal.ii would just share the pointers to the matrix data (or copy the data) with ginkgo and depending on what parameters you would give ginkgo, it would take care of the solution and return the solution vector in deal.ii's data structure.

**Comment by bangerth** [comment]
On 09/16/2018 02:36 PM, Pratik Nayak wrote:
> Yes, I understand that considering assembly time and the actual solution
> time, the time taken for copies would probably not make a difference. But
> as it is quite easy to prevent these copies by just declaring Ginkgo's
> sparse matrix class (a wrapper in deal.ii) a friend of the actual
> SparseMatrix I think it is doable. Of course, if you dont want outer
> libraries accessing the raw pointers or if you really prefer that we copy
> nonetheless, I am happy to do that as well.

No, let's not do this. Making other parts of (external) code a 'friend' is 
essentially as good as exposing the data structures right away -- it ties our 
hands to internal changes.

Let's just copy the matrix, if you don't mind.

**Comment by pratikvn** [comment]
Okay, no problem, I will do a copy.

**Comment by drwells** [comment]
Fixed by #7368.
